### PR TITLE
Revise SDL-0341 - Add Generic HMI Plugin Support (again)

### PR DESCRIPTION
**Pull request by jacobkeeler**
Revised proposal based on review comments (see https://github.com/smartdevicelink/sdl_evolution/issues/1168#issuecomment-964475414)

## Updates

- Add full definition for configuration file format
- Finalize inter-component message format (along with examples) and include API definitions for all RPCs
- Update wording throughout proposal to be more concrete, finalize several implementation details
- Clarify that all test plugins will be included directly in the repository

### CODE REVIEW

**Comment by jacobkeeler** [comment]
@theresalech Ready for review

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 1]
```diff
@@ -22,14 +22,40 @@ The main motivation behind this proposal is to bring the Generic HMI to a state
...
+* TTSPlugin (Plugin) - Configuration for the plugin which handles text-to-speech and connects to the `TTS` HMI interface. If enabled, the Generic HMI will not send `MB.registerComponent` for the `TTS` interface to the message broker.
+* NavigationPlugin (Plugin) - Configuration for the plugin which handles navigation-related functionality and connects to the `Navigation` HMI interface. This plugin does not override existing video/audio streaming functionality in the Generic HMI, as that is handled separately despite being defined in the `Navigation` interface.
+* MiscPlugins (Plugin array) - List of configurations for plugins are not tied to a specific HMI interface (such as the [General Settings](#general-settings-plugin) plugin or [Phone](#phone-plugin) plugin).
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:56]:
```suggestion
* MiscPlugins (Plugin array) - List of configurations for plugins which are not tied to a specific HMI interface (such as the [General Settings](#general-settings-plugin) plugin or [Phone](#phone-plugin) plugin).
```

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 2]
```diff
@@ -257,92 +302,320 @@ Connects to the `TTS` interface, overriding the Generic HMI's implementation of
...
+        Sent whenever a `Navigation.AlertManeuver` request is received by the Navigation plugin. The UI should display an alert maneuver popup with the provided softbuttons.
+    </description>
+    <param name="softButtons" type="Common.SoftButton" minsize="0" maxsize="3" array="true" mandatory="false">
+        <description>
+            Buttons to display with popup. 
+            If omitted, only the system defined "Close" SoftButton should be displayed.
+        </description>
+    </param>
+    <param name="appID" type="Integer" mandatory="true">
+        <description>ID of the application requested this RPC.</description>
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:460]:
```suggestion
        <description>ID of the application that requested this RPC.</description>
```

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 3]
```diff
@@ -257,92 +302,320 @@ Connects to the `TTS` interface, overriding the Generic HMI's implementation of
...
+<function name="ON_EVENT_CHANGED" messagetype="notification">
+    <description>
+        Direction: *Any Component* -> BasicCommunication
+        Sent when an event is activated or deactivated from a plugin to allow the main HMI to apply appropriate state changes.
+    </description>
+    <param name="eventName" type="EventTypes" mandatory="true">
+        <description>Specifies the type of event being activated or deactivated.</description>
+    </param>
+    <param name="isActive" type="Boolean" mandatory="true">
+       <description>'true' when the event is started, 'false' when the event has ended</description>
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:509]:
```suggestion
       <description>'true' when the event is started, 'false' when the event has ended.</description>
```

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 4]
```diff
@@ -257,92 +302,320 @@ Connects to the `TTS` interface, overriding the Generic HMI's implementation of
...
+<function name="DEACTIVATE_EVENT" messagetype="response">
+</function>
+```
+
+Additional Notes: The `destination` field for this request is dependent on the event type, according to which plugin is responsible for the event. Specifically, if the `eventName` is: 
+
+* `EMBEDDED_NAVIGATION` - the request will be sent to the `Navigation` interface.
+* `PHONE_CALL` - the request will be sent to the `BasicCommunication` interface (and handled by the Phone plugin).
+* `EMERGENCY_EVENT` - the request will be sent to the `BasicCommunication` interface (and handled by the General Settings plugin).
+* `DEACTIVATE_HMI` - the request will be sent to the `BasicCommunication` interface (and handled by the General Settings plugin.
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:542]:
```suggestion
* `DEACTIVATE_HMI` - the request will be sent to the `BasicCommunication` interface (and handled by the General Settings plugin).
```

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 5]
```diff
@@ -257,92 +302,320 @@ Connects to the `TTS` interface, overriding the Generic HMI's implementation of
...
+        Sent whenever a plugin needs to display a prompt to the user (ex. `RC.GetInteriorVehicleDataConsent` or `AppService.GetActiveServiceConsent`). A response is sent when an option is selected or the popup is closed.
+    </description>
+    <param name="title" type="String" mandatory="false">
+      <description>Title to display with the popup.</description>
+    </param>
+    <param name="popupText" type="String" mandatory="true">
+      <description>Text body of the popup.</description>
+    </param>
+    <param name="softButtons" type="SoftButton" mandatory="false" minsize="0" maxsize="4" array="true">
+      <description>SoftButtons to display with popup.</description>
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:565]:
```suggestion
      <description>SoftButtons to display with the popup.</description>
```

File: proposals/0341-add-generic-hmi-plugin-support.md [hunk 6]
```diff
@@ -257,92 +302,320 @@ Connects to the `TTS` interface, overriding the Generic HMI's implementation of
...
 
 ## Potential downsides
 
 Besides the general overhead of maintaining these plugins, the author does not see any significant downsides to this approach. Using plugins may be slightly more complicated than integrating these interfaces directly into the HMI, but it also allows for extra flexibility, and would not require any sizable design changes to the Generic HMI.
 
 ## Impact on existing code
 
-With the additional components being added outside of the main component of the HMI, it would likely be worthwhile to include a startup script in the Generic HMI project. This script could start the HMI, python backend, and any configured plugins all at once using a single command.
+With the variety of components being added in this proposal which run separate from the main HMI, a startup script will be required to handle the initialization of all of these components. This script will start the HMI, python backend, and any desired plugins all at once using a single command. This new script will need it's own simple configuration file in order to start each plugin, for example:
```
Comment (theresalech) [inline proposals/0341-add-generic-hmi-plugin-support.md:596]:
```suggestion
With the variety of components being added in this proposal which run separate from the main HMI, a startup script will be required to handle the initialization of all of these components. This script will start the HMI, python backend, and any desired plugins all at once using a single command. This new script will need its own simple configuration file in order to start each plugin, for example:
```

**Review by theresalech — CHANGES_REQUESTED** [review]
Hi @jacobkeeler, I've suggested a few minor changes. Please review and let me know if you have any questions or concerns with applying my feedback. Thanks!

File: proposals/0341-add-generic-hmi-plugin-support.md (hunk 2 above)
Comment (jacobkeeler) [reply proposals/0341-add-generic-hmi-plugin-support.md:460]:
```suggestion
        <description>ID of the application which requested this RPC.</description>
```

**Review by theresalech — APPROVED** [review]

**Pull request merged**